### DISCUSSION TITLE
GCompartmentBuilder ignores prefWidth/prefHeight with FreeformLayout

### DISCUSSION THREAD

**Discussion by rsoika** [eclipse-glsp/glsp#760] [Q&A] [answered]
I have a strange behavior when building a GCompartment using the `GCompartmentBuilder` with a FreeformLayout.

I follow the documentation for [Layout Container with freeform Layout](https://www.eclipse.org/glsp/documentation/clientlayouting/#layout-container).

My parent node uses a VBOX layout adding a Header with a label and add the container. 

My code is a direct adoption form the documentation and looks very simple like this :

```
    private GCompartment createContainerCompartmentTest(final PoolGNode node, final GDimension containerPrefSize) {
        Map<String, Object> layoutOptions = new HashMap<>();
        layoutOptions.put(H_GRAB, true);
        layoutOptions.put(V_GRAB, true);
        layoutOptions.put(GLayoutOptions.KEY_PREF_WIDTH, containerPrefSize.getWidth());
        layoutOptions.put(GLayoutOptions.KEY_PREF_HEIGHT, containerPrefSize.getHeight());

        return new GCompartmentBuilder(ModelTypes.CONTAINER) //
                .id(node.getId() + "_container") //
                .layout(GConstants.Layout.FREEFORM) //
                .layoutOptions(layoutOptions) //
                .build();
    }
```
This results in a 'rect' placed somehow randomly in my parent GNode:

![image](https://user-images.githubusercontent.com/1339124/193880456-6cd80b37-2285-4fd2-a2cf-0477643d89ba.png)

The computed width and height makes no sense at all as I set the values 800x200

At the end I can place nodes inside my Parent Node which looks fine on a first look. But When  a new element is placed in the area where the container compartment is placed I can no longer grap and move this node because the container rect blocks this.

Now, when I use the following Builder options where I do not set any layoutOptions like this:

```
    private GCompartment createContainerCompartment(final PoolGNode node, final GDimension containerPrefSize) {
        Map<String, Object> layoutOptions = new HashMap<>();
        return new GCompartmentBuilder(ModelTypes.CONTAINER) //
                .id(node.getId() + "_container") //
                .layout(GConstants.Layout.FREEFORM) //
                .layoutOptions(layoutOptions) //
                .build();
    }
```

everything is working great. The compartment rect has now a size of 0x0 and it does no longer block other childs:

![image](https://user-images.githubusercontent.com/1339124/193881311-088e1346-3bb4-42cd-ac7b-afcb7f3b9678.png)

I just want to ask about your opinion? For me it looks fine when the ContainerCompartment has no size. It is just used to hold the ChildList of GNode elements. 

In the moment I am very confused about the documentation which is  very clear  and the strange behavior of my final component.  

My requirement is also more complex as the parent node contains so called Lanes with a fixed position and size and also elements which can be moved around. This looks at the end like this:

![image](https://user-images.githubusercontent.com/1339124/193882919-842c19df-8a33-46de-b6b1-f6c2a80876dc.png)

**Comment by rsoika** [comment, marked as answer]
It looks like I found the solution. 
On the client side I defined the container model element as a `StructureCompartmentView` - as used in the workflow example:

    configureModelElement(context, 'container', SCompartment, StructureCompartmentView);

After I changed this to a simple `SCompartmentView` the 'rect' HTML node is gone:

    configureModelElement(context, 'container', SCompartment, SCompartmentView);

Everything seems now working as expected - I can move the container with its elements and the elements within the container.

**Comment by CamilleLetavernier** [comment]
Hi Ralph,

Another issue is that mixing hGrab/vGrab ("make the compartment match the size available in the parent, share with siblings if necessary") and "prefHeight/prefWidth" ("Use at least this size, but you can make it bigger if children require it") are not meant to be used together.

hGrab/vGrab delegates size management to the parent; prefHeight/prefWidth is intended for manually-resized nodes (Usually top-level elements). minSize would be more appropriate than prefSize in this case.

For swimlanes, custom layouts might be required, especially if you intend to manually resize the lanes. If not, then a vBox containing lanes defined with hGrab=true, vGrab=false (and maybe a minSize) should work (Then the actual size of the swimlane will depend on the children present inside it. This approach doesn't give much flexibility regarding margins, however).
